specifyweb/frontend/js_src/lib/components/WorkBench/WbSpreadsheet.tsx

```diff
@@ -0,0 +1,501 @@
+        manualColumnMove={true}
+        manualColumnResize={true}
+        minSpareRows={userPreferences.get(
+          'workBench',
+          'editor',
+          'minSpareRows'
+        )}
+        multiColumnSorting={true}
+        outsideClickDeselects={false}
+        placeholderCellClassName="htPlaceholder"
+        ref={hotRef}
+        rowHeaders={true}
+        // @ts-expect-error
+        sortIndicator={true}
+        stretchH="all"
+        tabMoves={
+          userPreferences.get('workBench', 'editor', 'tabMoveDirection') ===
+          'col'
+            ? { col: 1, row: 0 }
+            : { col: 0, row: 1 }
+        }
+        readOnly={isUploaded || !hasPermission('/workbench/dataset', 'update')}
+        contextMenu={contextMenuConfig as Settings}
+        {...hooks}
+      />
+      {noDisambiguationDialog && (
+        <Dialog
+          buttons={commonText.close()}
+          header={wbText.noDisambiguationResults()}
+          onClose={closeNoDisambiguationDialog}
+        >
+          {wbText.noDisambiguationResultsDescription()}
+        </Dialog>
+      )}
+      {disambiguationDialog && (
+        <DisambiguationDialog
+          matches={disambiguationResource!.models}
+          onClose={closeDisambiguation}
+          onSelected={(selected) => {
+            disambiguation.setDisambiguation(
+              disambiguationPhysicalRow!,
+              disambiguationMatches!.mappingPath,
+              selected.id
+            );
+            validation?.startValidateRow(disambiguationPhysicalRow!);
+            hotRef.current?.hotInstance!.render();
+          }}
+          onSelectedAll={(selected): void =>
+            // Loop backwards so the live validation will go from top to bottom
+            hotRef.current!.hotInstance!.batch(() => {
+              for (
+                let visualRow = data.length - 1;
+                visualRow >= 0;
+                visualRow--
+              ) {
+                const physicalRow =
+                  hotRef.current!.hotInstance!.toPhysicalRow(visualRow);
+                if (
+                  !validation?.uploadResults.ambiguousMatches[
+                    physicalRow
+                  ]?.find(
+                    ({ key, mappingPath }) =>
+                      key === disambiguationMatches!.key &&
+                      typeof disambiguation.getDisambiguation(physicalRow)[
+                        mappingPathToString(mappingPath)
+                      ] !== 'number'
+                  )
+                )
+                  continue;
+                disambiguation.setDisambiguation(
+                  physicalRow,
+                  disambiguationMatches!.mappingPath,
+                  selected.id
+                );
+                validation?.startValidateRow(physicalRow);
+              }
+            })
+          }
+        />
+      )}
+    </>
+  );
+}
+
+export const WbSpreadsheet = React.memo(WbSpreadsheetComponent);
```

Comment:
I know you spent quite a bit of time fighting the re-rendering issues here, but could you please refactor this to call useMemo() inside the component in addition to OR instead of `React.memo`?

call React.useMemo for each of the non-literal value props we pass to the Handsontable instance (`columns`, `contextMenuConfig` ... and anything else that is an object that is created on each render)

otherwise, handsontable might not like the fact that the `columns` array is re-created on each render...